internal/app/machined/pkg/controllers/runtime/kmsg_log.go

```diff
@@ -0,0 +1,182 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/AlekSi/pointer"
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/talos-systems/go-kmsg"
+	"github.com/talos-systems/go-procfs/procfs"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/talos-systems/talos/internal/app/machined/pkg/runtime"
+	"github.com/talos-systems/talos/internal/app/machined/pkg/runtime/logging"
+	"github.com/talos-systems/talos/pkg/machinery/constants"
+	"github.com/talos-systems/talos/pkg/machinery/resources/network"
+)
+
+const drainTimeout = 100 * time.Millisecond
+
+// KmsgLogDeliveryController watches events and forwards them to the events sink server
+// if it's configured.
+type KmsgLogDeliveryController struct {
+	Cmdline *procfs.Cmdline
+	Drainer *runtime.Drainer
+
+	drainSub *runtime.DrainSubscription
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *KmsgLogDeliveryController) Name() string {
+	return "runtime.KmsgLogDeliveryController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *KmsgLogDeliveryController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: network.NamespaceName,
+			Type:      network.StatusType,
+			ID:        pointer.ToString(network.StatusID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *KmsgLogDeliveryController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *KmsgLogDeliveryController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) (err error) {
+	if ctrl.Cmdline == nil || ctrl.Cmdline.Get(constants.KernelParamLoggingKernel).First() == nil {
+		return nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		var netStatus resource.Resource
+
+		netStatus, err = r.Get(ctx, resource.NewMetadata(network.NamespaceName, network.StatusType, network.StatusID, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				// no network state yet
+				continue
+			}
+
+			return fmt.Errorf("error reading network status: %w", err)
+		}
+
+		if !netStatus.(*network.Status).TypedSpec().AddressReady {
+			// wait for address
+			continue
+		}
+
+		break
+	}
+
+	if ctrl.drainSub == nil {
```

Comment:
I wonder if it makes sense to move drain sub closer to the loop to avoid blocking it when we return before entering the loop.

Reply:
or maybe it's fine. In the end it tries to reconnect there. So I guess it's fine.

Reply:
it subscribes for drains as soon as the address is ready, so it should be fine (I guess).